libs/llrb/llrb.go

```diff
@@ -0,0 +1,293 @@
+package llrb
+
+import (
+	"fmt"
+	"sync"
+	"time"
+)
+
+// MaxSize is the max allowed number of node a llrb is
+// allowed to contain.
+// If more nodes are pushed it will panic.
+const MaxSize = int(^uint(0) >> 1)
+
+type nodeKey struct {
+	priority uint64
+	ts       time.Time
+}
+
+func (a nodeKey) compare(b nodeKey) int {
+	if a.priority > b.priority {
+		return 1
+	}
+	if a.priority < b.priority {
+		return -1
+	}
+	if a.ts.Before(b.ts) {
+		return 1
+	}
+	if a.ts.After(b.ts) {
+		return -1
+	}
+	return 0
+}
+
+func NewNodeKey(priority uint64, ts time.Time) interface{} {
+	k := new(nodeKey)
+	k.priority = priority
+	k.ts = ts
+	return k
+}
+
+type node struct {
+	Key         nodeKey
+	Data        interface{}
+	Left, Right *node
+	Black       bool
+}
+
+type Llrb struct {
```

Comment:
for better mock-ability, how about defining an interface and let `llrb` implement that interface?